### PR TITLE
EventsWorker: Send HTTP results to the worker.

### PR DESCRIPTION
**Pull request by eloycoto**
Due to the nature of MQTT, the HTTP responses are not sent to the
worker(k4e-device-worker), so the worker cannot take decisions based on
the operator.

Like, renew keys to login into the operator. So a worker can check the
401 results to renew certificates.

With this change, the worker will have an Events function, that can
retrieve the Event information with the following information:

```
message APIResponse {
  int64 status_code = 1;

  string body = 2;

  string directive = 3;

  map<string, string> metadata = 4;
}
```

The events are triggered by the `transport.PublishReceivedData` to a
channel that it's async.

Signed-off-by: Eloy Coto <eloy.coto@acalustra.com>

### CODE REVIEW

File: cmd/yggd/grpc.go
Comment (eloycoto) [inline cmd/yggd/grpc.go:122]: When registering the worker, the key is "device", but here I'm not sure if we should range over *ALL* workers and send it the information.

File: protocol/yggdrasil.proto
Comment (eloycoto) [inline protocol/yggdrasil.proto:74]: Comments to be added here, still waiting for the first round of reviews.

File: cmd/yggd/grpc.go
Comment (pkliczewski) [reply cmd/yggd/grpc.go:122]: I think we need to

File: cmd/yggd/grpc.go
Comment (pkliczewski) [inline cmd/yggd/grpc.go:131]: Should we retry here?

File: errors.go
Comment (pkliczewski) [inline errors.go:31]: The type looks good but I am not sure whether it should be in this file. It feels correct to have it in `http.go`. What do you think?

File: internal/clients/http/http.go
Comment (pkliczewski) [inline internal/clients/http/http.go:103]: should this be `apiResponse, nil` instead?

**Review by pkliczewski — COMMENTED** [review]
We could add some tests

File: internal/clients/http/http.go
Comment (eloycoto) [reply internal/clients/http/http.go:103]: APIResponse implements the error() interface.

File: errors.go
Comment (eloycoto) [reply errors.go:31]: yup, to be moved.

**Comment by eloycoto** [comment]
Yes, test is a thing that I'll add. BUT want to know if it's the right way first.

**Comment by pkliczewski** [comment]
> Yes, test is a thing that I'll add. BUT want to know if it's the right way first.

Looks like good approach to me

File: cmd/yggd/grpc.go
Comment (eloycoto) [reply cmd/yggd/grpc.go:134]: I do not think so, workers can die at any moment, no? missing an event is not also critical here.

File: protocol/yggdrasil.proto
Comment (jakub-dzon) [inline protocol/yggdrasil.proto:20]: Please fix the comment.

File: cmd/yggd/grpc.go
Comment (jakub-dzon) [inline cmd/yggd/grpc.go:148]: My understanding of that code is that every event that is put on the eventQ (which is all the responses for all requests sent out to the control plane) is sent to every worker registered in yggdrasil - every request made by any worker causes N (number of workers)  events to be sent to every worker. 
Is that how that code works is there some implicit filtering/dispatching that I missed?

Not all workers are interested (or even should be) in all events - especially when the events are responses to requests. That also might be considered a security hole - exposing information directed to only one recipient to others. 

Only worker sending request that causes the event should receive the event.

File: internal/transport/http/http.go
Comment (jakub-dzon) [inline internal/transport/http/http.go:49]: `APIResponse` should be lowercase.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]
There is "almost merged" [PR](https://github.com/RedHatInsights/yggdrasil/pull/39) to upstream yggdrasil that will significantly affect this PR.

Please consult the changes with Link, as they modify yggdrasil internals and API significantly.

File: internal/transport/publising.go
Comment (masayag) [inline internal/transport/publising.go:69]: this condition reduces the response for the http transport only, right ? since MQTT.SendData is expected to return nil.
Could this be encapsulated as part of the http.SendData if this behavior reserved for a specific type of transport ?

File: protocol/yggdrasil.proto
Comment (masayag) [reply protocol/yggdrasil.proto:74]: The API looks good to me. I suggest to get Link's review on this, since he'll have to approve the API when it is contributed back to main branch.

**Pull request closed without merging**
